resources/default.rb

```diff
@@ -18,32 +18,691 @@
 # If the upgrade is scheduled in the same minute that Chef runs,
 # there is a risk the upgrade will not be triggered.
 # Lowering upgrade_delay limit is not recommended.
-attribute :upgrade_delay, kind_of: Integer, default: 60
-attribute :product_name, kind_of: String, default: 'chef'
-attribute :rubygems_url, kind_of: String, default: lazy { Chef::Config[:rubygems_url] }
-attribute :handle_zip_download_url, kind_of: String, default: 'https://download.sysinternals.com/files/Handle.zip'
-attribute :event_log_service_restart, kind_of: [TrueClass, FalseClass], default: true
-attribute :install_command_options, kind_of: Hash, default: {}
+property :upgrade_delay, Integer, default: 60
+property :product_name, String, default: 'chef'
+property :rubygems_url, String, default: lazy { Chef::Config[:rubygems_url] }
+property :handle_zip_download_url, String, default: 'https://download.sysinternals.com/files/Handle.zip'
+property :event_log_service_restart, [true, false], default: true
+property :install_command_options, Hash, default: {}
+
+action_class do
+  def load_mixlib_install
+    gem 'mixlib-install', '~> 3.12'
+    require 'mixlib/install'
+  rescue LoadError
+    Chef::Log.info('mixlib-install gem not found. Installing now')
+    chef_gem 'mixlib-install' do
+      version '~> 3.12'
+      compile_time true if respond_to?(:compile_time) # cookstyle: disable Chef/Modernize/RespondToCompileTime
+      if new_resource.rubygems_url
+        clear_sources true if respond_to?(:clear_sources)
+        if respond_to?(:source)
+          source new_resource.rubygems_url
+        elsif respond_to?(:options)
+          options "--source #{new_resource.rubygems_url}"
+        end
+      end
+    end
+
+    require 'mixlib/install'
+  end
+
+  def load_mixlib_versioning
+    require 'mixlib/versioning'
+  rescue LoadError
+    Chef::Log.info('mixlib-versioning gem not found. Installing now')
+    chef_gem 'mixlib-versioning' do
+      compile_time true if respond_to?(:compile_time) # cookstyle: disable Chef/Modernize/RespondToCompileTime
+      if new_resource.rubygems_url
+        clear_sources true if respond_to?(:clear_sources)
+        options "--source #{new_resource.rubygems_url}" if respond_to?(:options)
+      end
+    end
+
+    require 'mixlib/versioning'
+  end
+
+  def update_rubygems
+    compatible_rubygems_versions = '>= 2.6.11'
+    target_version = '2.7.8' # should be bumped to latest 2.x, but ruby < 2.3.0 support is necessary
+    nodoc_rubygems_versions = '>= 3.0'
+
+    rubygems_version = Gem::Version.new(Gem::VERSION)
+    Chef::Log.debug("Found gem version #{rubygems_version}. Desired version is #{compatible_rubygems_versions}")
+    return if Gem::Requirement.new(compatible_rubygems_versions).satisfied_by?(rubygems_version)
+
+    # only rubygems >= 1.5.2 supports pinning, and we might be coming from older versions
+    pin_rubygems_range = '>= 1.5.2'
+    pin = Gem::Requirement.new(pin_rubygems_range).satisfied_by?(rubygems_version)
```

Comment:
Chef 12.5 comes with 2.4.8 so pinning is always supported now. We can remove all this logic